src/machine.cpp

```diff
@@ -6157,6 +6201,68 @@ GcCallSite* resolveDynamic(Thread* t, GcInvocation* invocation)
     ++i;
   }
 
+  object varArgsArray_ = makeObjectArray(t, numVarArgs);
+  PROTECT(t, varArgsArray_);
+
+  if (numVarArgs > 0) {
+    assertT(t, i < bootstrapArray->length());
+    GcArray *varArgsArray = reinterpret_cast<GcArray*>(varArgsArray_);
```

Comment:
Awesome, I'd love to read that!

Yeah, PROTECT is definitely fickle.  It's also one of those things whose misuse to the damnedest bugs.

They key insight is that we don't trigger a GC unless we're actually trying to allocate something -  which means there's no way for a GC to happen between creating that variable and calling PROTECT.  It's been a while since I worked in that code, but IIRC, there's something preventing us from considering a thread running VM code as being at a "safe-point", so other threads also can't trigger a GC.

In general, having an extra PROTECT call is a tiny performance problem, but missing one leads to things like memory corruption when the object gets moved but the pointer isn't updated.  So, it's better to err on the side of caution.

There's a PROTECT call in the generated `makeWhatever` constructors, so you can allocate one object and pass that to the constructor of another object, but you can't pass multiple unPROTECTed objects to a constructor, since constructing the second one of those might trigger a GC.

As a side-note, I've been a huge fan of Rust (the language) for a while - and one of the features that's under discussion right now is [enabling GC integration in the compiler](http://manishearth.github.io/blog/2016/08/18/gc-support-in-rust-api-design/), which would make it dramatically easier to write GC'd language runtimes in rust.  Not that there's any plan to translate avian into rust.  Honest.